<a href="https://colab.research.google.com/github/eisakhan04/Numpy-Course/blob/main/Topic_08_%3D_NumPy_Linear_Algebra.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Topic 08 = NumPy Linear Algebra**
NumPy is a powerful library in Python used for scientific computing, especially for working with arrays and matrices. One of the key features of NumPy is its support for linear algebra operations.

Linear algebra is a branch of mathematics that deals with linear equations, matrices, and vectors. In NumPy, linear algebra operations are performed using the linalg module. This module contains a wide range of functions for performing various linear algebra operations, such as matrix multiplication, solving linear equations, finding determinants and eigenvalues, and performing singular value decomposition (SVD).

Using NumPy's linear algebra capabilities, you can efficiently and accurately perform complex calculations on matrices and arrays, making it a valuable tool for scientific computing and data analysis.

Numpy provides the following functions to perform the different algebraic calculations on the input data.

**(SN): (Function): (Definition):**

1. dot() It is used to calculate the dot product of two arrays.

2. vdot() It is used to calculate the dot product of two vectors.

3. inner() It is used to calculate the inner product of two arrays.

4. matmul() It is used to calculate the matrix multiplication of two arrays.

5. det() It is used to calculate the determinant of a matrix.

6. solve() It is used to solve the linear matrix equation.

7. inv() It is used to calculate the multiplicative inverse of the matrix.

**1.numpy.dot() function**

The **numpy.dot()** function is a part of NumPy's linear algebra module that is used to perform matrix multiplication between two arrays. It takes two arrays as input and returns their dot product.

The dot product of two arrays is calculated by multiplying the corresponding elements of the arrays and then summing them up. The dot product of two arrays A and B can be denoted by A.B or A*B.

The **numpy.dot()** function can be used with arrays of different shapes, as long as the dimensions of the arrays are compatible for matrix multiplication. Specifically, the number of columns in the first array must be equal to the number of rows in the second array. If the shapes of the arrays are not compatible for matrix multiplication, the function will raise a **ValueError**.

**Example 01**




In [ ]:
import numpy as np
a = np.array([[100,200],[23,12]])
b = np.array([[10,20],[12,21]])
print(a)
print()
print(b)
print()
dot = np.dot(a,b)
print( "Values of dot product:", dot)


[[100 200]
 [ 23  12]]

[[10 20]
 [12 21]]

Values of dot product: [[3400 6200]
 [ 374  712]]



**Explanation**

This code defines two 2D NumPy arrays a and b, each with 2 rows and 2 columns. It then prints both arrays to the console using the print() function.

Next, the code uses the numpy.dot() function to perform matrix multiplication between a and b. The result is stored in a variable named mul.

Finally, the code prints a message to the console along with the value of mul using the print() function.

The values of the resulting array mul are calculated as follows:

mul[0,0] = (100 X 10) + (200 X 12) = 3400

mul[0,1] = (100 X 20) + (200 X 21) = 6200

mul[1,0] = (23 X 10) + (12 X 12) = 374

mul[1,1] = (23 X 20) + (12 X 21) = 712

Therefore, the output of print("Values of dot product:", dot ) will be:

matrix multiplication: [[3400 6200] [ 374 712]]

In [ ]:
# Example 02


import numpy as np
a = np.array([100,200])
b = np.array([23,12])
dot = np.dot(a,b)
print(dot)


4700


     
**Explanation**

In this example, the array function is used to create two 1D arrays a and b, each with two elements. The values in the arrays are specified using a list.

The numpy.dot() function is then used to perform dot product between arrays a and b, and the resulting dot product is stored in the dot variable.

Finally, the dot product value is printed to the console using the print() function.

So, This is the result of taking the dot product of arrays a and b using the numpy.dot() function. The dot product of two 1D arrays is simply the sum of the element-wise product of the two arrays.

**2.numpy.vdot() function**

The numpy.vdot() function is a NumPy library function that computes the dot product of two 1D arrays. It returns the dot product of the two input arrays, which is computed as the sum of the element-wise product of the two arrays.

Unlike the numpy.dot() function, which can be used to compute the dot product of arrays of any dimension, the numpy.vdot() function is specifically designed to work with 1D arrays only.

Here's the syntax for using the numpy.vdot() function:

numpy.vdot(a, b)

**Example 01**




In [ ]:
import numpy as np
a = np.array([[100,200],[23,12]])
print(a)
b = np.array([[10,20],[12,21]])
print(b)
print()
vdot = np.vdot(a,b)
print("values of vector dot product:", vdot)

[[100 200]
 [ 23  12]]
[[10 20]
 [12 21]]

values of vector dot product: 5528


Explanation

In this example, the vector dot product of the matrices a and b is calculated as follows:

a = [100, 200, 23, 12]

b = [10, 20, 12, 21]

vdot = a[0] X b[0] + a[1] X b[1] + a[2] X b[2] + a[3] X b[3]

 = 100 X 10 + 200 X 20 + 23 X 12 + 12 X 21

 = 5528



In [ ]:
# Example 02


import numpy as np

a = np.array([1 + 2j, 3 + 4j])
b = np.array([5 + 6j, 7 + 8j])
vdot = np.vdot(a, b)
print(vdot)


(70-8j)


**Explanation**

Two complex-valued vectors, a and b, are defined using the np.array() function with the values [1 + 2j, 3 + 4j] and [5 + 6j, 7 + 8j], respectively.

The np.vdot() function is then used to calculate the vector dot product of the two vectors a and b. The vector dot product of two vectors a and b is defined as the sum of the element-wise products of the conjugate of the elements of a and the corresponding elements of b. In other words, it is equivalent to taking the dot product of a with the conjugate of b.

The result of the vector dot product is stored in the variable vdot, and printed to the console using the print() function.

In this example, the vector dot product of the complex vectors a and b is calculated as follows:

vdot = (1 - 2j) X (5 + 6j) + (3 - 4j) X (7 + 8j)

 = (5 + 6j - 10j + 12) + (21 + 24j - 28j + 32)

 = 17 - 4j + 53 -4j

 = 70 - 8j

**3.numpy.inner() function**

The numpy.inner() function is a part of the NumPy library in Python, which is used for scientific computing. This function is used to calculate the inner product of two arrays.

The inner product of two arrays a and b is defined as the sum of the product of corresponding elements in both arrays. It can also be calculated by multiplying the transpose of a with b or vice versa.

**The syntax for using the numpy.inner() function is:**

numpy.inner(a, b)

where a and b are the input arrays. The function returns the inner product of the two arrays as a scalar value.

If a and b are both 1-D arrays, numpy.inner() calculates the dot product of the two arrays. If either of the arrays is multi-dimensional, the product sum will be over the last dimension of the first array and the second-to-last dimension of the second array.

The numpy.inner() function can also be used to calculate the inner product of complex arrays. In this case, the complex conjugate of the first array is used.

**Example 01**




In [ ]:
import numpy as np
a = np.array([1,2,3,4,5,6])
b = np.array([23,23,12,2,1,2])
inner = np.inner(a,b)
print(inner)

130


**Explanation**

This code defines two NumPy arrays a and b containing six elements each. Then, it uses the numpy.inner() function to calculate the inner product of a and b. The result is stored in a variable named inner.

Finally, the code prints the value of inner to the console using the print() function.

The output of this code will be a single scalar value, which is the inner product of a and b. This value represents the sum of the product of corresponding elements in both arrays. In this case, it will be:

(1 X 23) + (2 X 23) + (3 X 12) + (4 X 2) + (5 X 1) + (6 X 2) = 130

Therefore, the output of print(inner) will be 130.

**Example 02**



**Explanation**

This code defines two NumPy arrays a and b, each with 2 rows and 3 columns. It then prints both arrays to the console using the print() function.

Next, the code uses the numpy.inner() function to calculate the inner product of a and b. The result is stored in a variable named inner.

Finally, the code prints a message to the console along with the value of inner using the print() function.

Since a and b are both 2D arrays, the numpy.inner() function calculates the inner product along the last dimension of the first array and the second-to-last dimension of the second array. In this case, the inner product is calculated for each row of a and each column of b, resulting in a 2x2 array.

The values of the resulting array inner are calculated as follows:

inner[0,0] = (1 X 23) + (2 X 23) + (3 X 12) = 105

inner[0,1] = (1 X 2) + (2 X 1) + (3 X 2) = 10

inner[1,0] = (4 X 23) + (5 X 23) + (6 X 12) = 279

inner[1,1] = (4 X 2) + (5 X 1) + (6 X 2) = 25

In [ ]:
import numpy as np
a = np.array([[1,2,3],[4,5,6]])
b = np.array([[23,23,12],[2,1,2]])
print(a)
print()
print(b)
print()
inner = np.inner(a,b)
print(" Values of inner product:",inner)


[[1 2 3]
 [4 5 6]]

[[23 23 12]
 [ 2  1  2]]

 Values of inner product: [[105  10]
 [279  25]]



**4. numpy.matmul() function**

The numpy.matmul() function is a part of the NumPy library in Python, which is used for scientific computing. This function is used to perform matrix multiplication between two arrays.

The syntax for using the numpy.matmul() function is:

numpy.matmul(a, b)

where a and b are the input arrays. The function returns the matrix product of the two arrays.

The arrays a and b must have compatible shapes for matrix multiplication. Specifically, the number of columns in a must be equal to the number of rows in b.

If both a and b are 2-D arrays, then numpy.matmul() gives the same result as the @ operator or the numpy.dot() function.

The numpy.matmul() function can also be used to multiply higher-dimensional arrays, as long as the last dimension of a matches the second-to-last dimension of b.

If either a or b is a scalar (i.e., a single number), it is treated as a 1-D array.

It's important to note that numpy.matmul() is different from numpy.multiply(), which performs element-wise multiplication between two arrays.

**Example 01 :Matrix product**


In [ ]:
import numpy as np
a = np.array([[100,200],[23,12]])

b = np.array([[10,20],[12,21]])
print(a)
print()
print(b)
print()
mul = np.matmul(a,b)
print("matrix multiplication:", mul)

[[100 200]
 [ 23  12]]

[[10 20]
 [12 21]]

matrix multiplication: [[3400 6200]
 [ 374  712]]


**Explanation**

This code defines two 2D NumPy arrays a and b, each with 2 rows and 2 columns. It then prints both arrays to the console using the print() function.

Next, the code uses the numpy.matmul() function to perform matrix multiplication between a and b. The result is stored in a variable named mul.

Finally, the code prints a message to the console along with the value of mul using the print() function.

The values of the resulting array mul are calculated as follows:

mul[0,0] = (100 X 10) + (200 X 12) = 3400

mul[0,1] = (100 X 20) + (200 X 21) = 6200

mul[1,0] = (23 X 10) + (12 X 12) = 374

mul[1,1] = (23 X 20) + (12 X 21) = 712

Therefore, the output of print("matrix multiplication:", mul) will be:

matrix multiplication: [[3400 6200] [ 374 712]]

**Example 02 Element-wise matrix multiplication**



     
**Explanation**

The code defines two matrices using NumPy library in Python. The first matrix 'a' contains the values 1, 2, 3 in the first row, 4, 5, 6 in the second row, and 7, 8, 9 in the third row. The second matrix 'b' contains the values 9, 8, 7 in the first row, 6, 5, 4 in the second row, and 3, 2, 1 in the third row.

The code then multiplies these two matrices element-wise using the NumPy function 'np.multiply()' and stores the result in a new matrix 'c'. The resulting matrix 'c' contains the product of the corresponding elements of matrices 'a' and 'b'.

Finally, the code prints out the three matrices - 'a', 'b', and 'c' - using the 'print()' function to display the original matrices and the result of their element-wise multiplication.

In [ ]:
import numpy as np
a=np.array([[1,2,3],[4,5,6],[7,8,9]])
print("Array 1:",a)
print()
b=np.array([[9,8,7],[6,5,4],[3,2,1]])
print("Array 2:",b)
print()
c=np.multiply(a,b)
print("Matrix Multiplication:",c)


Array 1: [[1 2 3]
 [4 5 6]
 [7 8 9]]

Array 2: [[9 8 7]
 [6 5 4]
 [3 2 1]]

Matrix Multiplication: [[ 9 16 21]
 [24 25 24]
 [21 16  9]]


**5.numpy determinant**

Numpy is a popular Python library used for scientific computing and data analysis. It provides a variety of mathematical functions and operations, including calculating the determinant of a matrix.

The determinant of a matrix is a scalar value that can be computed from the elements of the matrix. It is commonly used in linear algebra and has several applications, such as solving systems of linear equations and calculating eigenvalues and eigenvectors.

In numpy, the determinant of a matrix can be calculated using the numpy.linalg.det() function. This function takes a matrix as an argument and returns its determinant as a single scalar value.

**Example**



**Explanation**

The code snippet you provided calculates the determinant of a 2x2 matrix using the numpy library in Python.

The matrix used in this example is defined as follows:

a = np.array([[1,2],[3,4]])

The determinant of this matrix is calculated using the np.linalg.det() function, which is a built-in function in numpy for computing the determinant of a matrix. The function takes the matrix as an argument and returns the determinant as a single value.

In [ ]:
import numpy as np
a = np.array([[1,2],[3,4]])
print(np.linalg.det(a))


-2.0000000000000004


**6.numpy.linalg.solve() function**

Numpy is a popular Python library used for scientific computing and data analysis. It provides a variety of mathematical functions and operations, including solving systems of linear equations.

The numpy.linalg.solve() function is used to solve a system of linear equations of the form Ax = b, where A is a square matrix, x is a column vector, and b is a column vector. The function returns the solution vector x that satisfies the equation.

Note that the numpy.linalg.solve() function can only be used to solve systems of linear equations with a square coefficient matrix. If the coefficient matrix is not square, the function will raise a LinAlgError exception.

**Example**




In [ ]:
import numpy as np
a = np.array([[3,2],[1,1]])
b = np.array([3,2])
print(a)
print(b)
print()
print(np.linalg.solve(a, b))

[[3 2]
 [1 1]]
[3 2]

[-1.  3.]


**Explanation**

This code snippet is written in Python and uses the numpy library to solve a system of linear equations.

First, a 2x2 matrix a is defined using the np.array() function with the values [3, 2] and [1, 1]. Then, a column vector b is defined using the np.array() function with the values [3, 2].

Next, the matrix a and vector b are printed to the console using the print() function. The print() function is called with an empty () after print(b) to add a blank line for clarity.

Finally, the np.linalg.solve() function is used to solve the system of linear equations Ax=b, where A is the matrix a and b is the vector b. The solution vector x is stored in the variable x and printed to the console using the print() function.

In this example, the np.linalg.solve() function is used to solve the system of linear equations 3x + 2y = 3 and x + y = 2. The solution vector x is [-1, 3], which satisfies both equations.



**7. numpy.linalg.inv() function**

Numpy is a popular Python library used for scientific computing and data analysis. It provides a variety of mathematical functions and operations, including matrix inversion.

The numpy.linalg.inv() function is used to calculate the inverse of a square matrix. The inverse of a matrix is a matrix that, when multiplied by the original matrix, produces the identity matrix.

Here's an example of how to use the numpy.linalg.inv() function to calculate the inverse of a matrix:

Note that the numpy.linalg.inv() function can only be used to calculate the inverse of square matrices that are invertible (i.e., matrices with a nonzero determinant). If the matrix is not invertible, the function will raise a LinAlgError exception.

**Example**



**Explanation**

The code snippet you provided is calculating the inverse of a 2x2 matrix using the numpy library in Python.

The matrix used in this example is defined as follows:

a = np.array([[1,2],[3,4]])

The inverse of this matrix is calculated using the np.linalg.inv() function, which is a built-in function in numpy for computing the inverse of a matrix. The function takes the matrix as an argument and returns the inverse matrix as a numpy array.

In this case, the inverse of the matrix 'a' is:

[[-2. , 1. ], [ 1.5, -0.5]]

The result is a numpy array containing the values of the inverse matrix. The code also prints the original matrix 'a' and the inverse matrix 'b' for comparison purposes.

Note that not all matrices have an inverse. A matrix must be square and non-singular (i.e., its determinant must be non-zero) to have an inverse.

In [ ]:
import numpy as np
a = np.array([[1,2],[3,4]])
print("Original array:\n",a)
b = np.linalg.inv(a)
print("Inverse:\n",b)


Original array:
 [[1 2]
 [3 4]]
Inverse:
 [[-2.   1. ]
 [ 1.5 -0.5]]
